json-streaming/src/main/scala/akka/stream/alpakka/json/JsonStreamReader.scala

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2016-2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.json
+
+import akka.annotation.InternalApi
+
+import scala.collection.immutable.Queue
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import akka.stream._
+import akka.util.ByteString
+import org.jsfr.json.exception.JsonSurfingException
+import org.jsfr.json.path.JsonPath
+import org.jsfr.json.{JsonPathListener, JsonSurferJackson, ParsingContext}
+
+/**
+ * Internal API
+ */
+@InternalApi
+private[akka] final class JsonStreamReader(path: JsonPath) extends GraphStage[FlowShape[ByteString, ByteString]] {
+
+  private val in = Inlet[ByteString]("Json.in")
+  private val out = Outlet[ByteString]("Json.out")
+  override val shape = FlowShape(in, out)
+
+  override def initialAttributes: Attributes = Attributes.name(s"jsonReader($path)")
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      private val in = shape.in
+      private val out = shape.out
+      setHandlers(in, out, this)
+
+      private var buffer = Queue.empty[ByteString]
+
+      private val surfer = JsonSurferJackson.INSTANCE
+      private val config = surfer.configBuilder
+        .bind(path, new JsonPathListener {
+          override def onValue(value: Any, context: ParsingContext): Unit =
+            buffer = buffer.enqueue(ByteString(value.toString))
```

Comment:
`buffer` is a var and is changed from the `onPush` handler. Therefore changing it here as well is not thread-safe. You should create an async callback that enqueues an element to the buffer and invoke that callback from the `JsonPathListener.onValue`.

Reply:
Will `onPush` ever be called concurrently? I know it's awkward, but the interface itself seems to be synchronous, as in: The handler was already called after `feed` on the parser returns. I'd have to doublecheck this via source-code inspection but my experiments have shown it to work this way until now.

If that's the case, we'd be fine wrt. thread-safety, right?

Reply:
`onPush` is not called concurrently. But the `buffer` var is modified from `onPush` and then from `onValue` which is a callback from jsonsurfer.

If the jsonsurfer does not have its own threads and the `onValue` callback is called by the same thread that called `parser.feed` in `onPush` then the current implementation is fine.

Reply:
I just checked and the callback is called from the same thread as the `parser.feed` is called. Therefore this is good and does not need any modifications.